src/common/compat.cc

```diff
@@ -0,0 +1,78 @@
+#include <fcntl.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <sys/param.h>
+#include <sys/mount.h>
+#if defined(__linux__) 
+#include <sys/vfs.h>
+#endif
+
+#include "include/compat.h" 
+
+// The type-value for a ZFS FS in fstatfs.
+#define FS_ZFS_TYPE 0xde
+
+// On FreeBSD, ZFS fallocate always fails since it is considered impossible to
+// reserve space on a COW filesystem. posix_fallocate() returns EINVAL
+// Linux in this case already emulates the reservation in glibc
+// In which case it is allocated manually, and still that is not a real guarantee
+// that a full buffer is allocated on disk, since it could be compressed.
+// To prevent this the written buffer needs to be loaded with random data.
+
+int manual_fallocate(int fd, off_t offset, off_t len) {
+    // Try to manually allocate the buffer
+    char data[1024*128];
```

Comment:
move the definition of `data` to where it is used. also, might want pre-populate data with something. like 
```c
// populate data with random bits.
memset(data, 0x42, sizeof(data));
```
otherwise the static analyzer will be complaining.

Reply:
@tchaikov 
I was also contemplation in filling it with random data.
In that sense it would really reserve space on a compressing ZFS partition. Otherwise 1024^2 baytes will be reduced to almost nothing, and it is nor really a reservation.
But I'll start with this